Реализация паттерна MVVM.

PitBul477 opened the pull request:
Текстовое поле с вводом города заменено на ComboBox с выбором из заранее подготовленного выбора

Commit:
Реализация паттерна MVVM.
Текстовое поле с вводом города заменено на ComboBox с выбором из заранее подготовленного выбора

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,57 @@
+        private City selectedCity;
+
+        //Свойство для динамической коллекции городов
+        public ObservableCollection<City> Citys { get; set; }
```
Cities
+ set не нужен, если ты кроме конструктора нигде не назначаешь

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,57 @@
+                new City {CityName = "Осака", ID = 4 },
+                new City {CityName = "Квебек", ID = 5 }
+            };
+            _GetInfoWeather = new RelayCommand((obj) => { selectedCity.URL(); });
```
obj не нужен

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,57 @@
+            _GetInfoWeather = new RelayCommand((obj) => { selectedCity.URL(); });
+        }
+
+        private RelayCommand _GetInfoWeather;
```
это можно убрать, оставь просто GetInfoWeather.
Выделять поля для свойств нужно если ты что то принимаешь через конструктор и это надо сохранить.

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,57 @@
+                new City {CityName = "Череповец", ID = 2 },
+                new City {CityName = "Токио", ID = 3 },
+                new City {CityName = "Осака", ID = 4 },
+                new City {CityName = "Квебек", ID = 5 }
```
ID не нужен

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,57 @@
+        //конструктор модели представления, создаёт объекты для коллекции городов и задаёт команду
+        public ViewModels()
+        {
+            Citys = new ObservableCollection<City>
```
ObservableCollection нужно использовать если ты будешь менять коллекцию в реальном времени.
тут достаточно IEnumerable<City>

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,39 @@
+            remove { CommandManager.RequerySuggested -= value; }
+        }
+
+        public RelayCommand(Action<object> execute, Func<object, bool> canExecute = null)
```
Достаточно просто Action

LilySweetCat replied:
Комментарии по всему классу

LilySweetCat commented on Task1/View/MainWindow.xaml.cs:
```diff
@@ -0,0 +1,21 @@
+        public MainWindow()
+        {
+            InitializeComponent();
+            DataContext = new ViewModels();
```
Можно через xaml.
<Window.DataContext>
        <local:MainViewModel/>
    </Window.DataContext>
Конструктор VM должен быть публичным

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -1,12 +1,12 @@
 	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
 	xmlns:properties="clr-namespace:Task1.Properties"
-	Title="MainWindow"
+	Title="Прогноз погоды"
```
хм... А может в ресурсы?

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -20,26 +20,33 @@
 				<ColumnDefinition Width="60*" />
```
0.6* и 0.4*

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -20,26 +20,33 @@
+				DisplayMemberPath="CityName"
+				ItemsSource="{Binding Path=Citys}"
+				SelectedItem="{Binding Path=SelectedCity}"
+				SelectedValuePath="ID" />
```
не нужен

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -20,26 +20,33 @@
 				Grid.Column="0"
-				Margin="20,12,10,12" />
+				Margin="20,12,10,12"
+				DisplayMemberPath="CityName"
```
сделай через item template

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -20,26 +20,33 @@
-				Margin="20,12,10,12" />
+				Margin="20,12,10,12"
+				DisplayMemberPath="CityName"
+				ItemsSource="{Binding Path=Citys}"
```
Cities

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -20,26 +20,33 @@
 				Margin="10"
-				Click="Button_Click"
+				Command="{Binding GetInfoWeather}"
+				CommandParameter="{Binding SelectedCity}"
```
убери, пусть VM все делает сама

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        //Свойство для поля id
+        public int ID
```
не нужен

Antonybae commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+        public string CityUrl
+        {
+            get { return cityUrl; }
+            set
```
приватный т.к. извне ты ничего не меняешь

Antonybae replied:
!

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,39 @@
+    public class RelayCommand : ICommand
+    {
+        private Action<object> execute;
+        private Func<object, bool> canExecute;
```
замени на Predicate

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+            set
+            {
+                cityUrl = value;
+                OnPropertChanged("");
```
почему ""?

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        //Функция вычисления cityUrl из других полей и с помощью get-запроса
+        public void URL()
```
GetURL

LilySweetCat replied:
По правилам методы должны начинаться с глагола

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+    {
+        const string SiteUrl = "http://weather.service.msn.com/data.aspx?weasearchstr=";
+        const string ParamUrl = "&culture=en-US&weadegreetype=C&src=outlook";
+        private string cityName = "";
```
string.Empty
ниже тоже

LilySweetCat replied:
string.Empty
ниже тоже

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+        const string SiteUrl = "http://weather.service.msn.com/data.aspx?weasearchstr=";
+        const string ParamUrl = "&culture=en-US&weadegreetype=C&src=outlook";
+        private string cityName = "";
+        static string cityUrl = "";
```
смысл в статике?
вдруг я захочу поменять чекбоксом выбранный город?

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,79 @@
+            set
+            {
+                cityName = value;
+                OnPropertChanged("");
```
""

LilySweetCat reviewed (commented):
правь ^_^

Commit:
Мелкие исправления

Commit:
Исправление недочётов:
1) DisplayMemberPath заменён на ItemTemplate.
2) Изменена реализация класса RelayCommand.
3) ObservableCollection заменён на List (как реализация IEnumerable).
И другое

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,67 @@
+    //Класс модели City
+    public class City : INotifyPropertyChanged
+    {
+        const string SiteUrl = "http://weather.service.msn.com/data.aspx?weasearchstr=";
```
private
и ниже тоже

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,67 @@
+        private string cityName = string.Empty;
+        private string cityUrl = string.Empty;
+
+        //Свойство для поля cityName
```
комментарий для свойств должен быть как коммент для полей т.е. полным
в данном случае "Имя города"

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,67 @@
+            }
+        }
+
+        //Свойство для поля cityUrl
```
тоже

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        //событие изменения свойства
+        public event PropertyChangedEventHandler PropertyChanged;
```
добавь отступ между
это поле перенеси к остальным полям
сверху должны быть поля, потом свойства потом конструктор, потом публичные методы потом приватные

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,67 @@
+        //Функция вычисления cityUrl из других полей и с помощью get-запроса
+        public void GetUrl()
+        {
+            var url = ($"{SiteUrl}{CityName}{ParamUrl}");
```
лишние скобки

LilySweetCat commented on Task1/View/MainWindow.xaml:
```diff
@@ -1,45 +1,61 @@
+				SelectedItem="{Binding Path=SelectedCity}">
+				<ComboBox.ItemTemplate>
+					<DataTemplate>
+						<StackPanel>
```
достаточно просто textblock

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,45 @@
+        //конструктор класса RelayCommand
+        public RelayCommand(Action execute, Predicate<object> canExecute = null)
+        {
+            this.execute = execute;
```
проверку на null

LilySweetCat replied:
если null, то кинь исключение типа Argument

LilySweetCat reviewed (commented):
ты на верном пути

Commit:
Исправление ошибок

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,70 @@
+        //Функция вычисления cityUrl из других полей и с помощью get-запроса
+        public void GetUrl()
+        {
+            if (CityName == null)
```
это когда это у тебя вообще может быть CityName = null?

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,66 @@
+                OnPropertyChanged("selectedCity");
+            }
+        }
+
```
что за куча отступов?

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,66 @@
+        public ViewModels()
+        {
+            Cities = new List<City> {
+                new City {CityName = "Москва" },
```
передавай через конструктор!!!

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,66 @@
+            GetInfoWeather = new RelayCommand(() => { SelectedCity.GetUrl(); }, Pder);
+        }
+
+        private bool Pder(object obj)
```
что за название XD
дай нормальное
раз возвращает bool, то должно начинаться на Is или Check

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,70 @@
+            set
+            {
+                _cityName = value;
+                OnPropertChanged(string.Empty);
```
убери OnPropertyChanged и сделай getter приватным.
предварительно не забудь сделать конструктор!!!

LilySweetCat replied:
в конструкторе проверь на null

LilySweetCat replied:
если у свойства есть get, то комментарий должен начинаться на Возвращает
если есть set, то Устанавливает
если и то и то, то Возвращает или устанавливает

LilySweetCat reviewed (commented):
лол

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        //команда на запуск формирования get-запроса
+        public RelayCommand GetInfoWeather { get; }
```
возвращает

Antonybae commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,66 @@
+        public IEnumerable<City> Cities { get; }
+
+        //свойство для поля selectedCity, которое хранит в себе выбранный в ComboBox-е город
+        public City SelectedCity
```
возвращает или устанавливает
ну ты понял

LilySweetCat commented on Task1/ViewModel/ViewModels.cs:
```diff
@@ -0,0 +1,66 @@
+    public class ViewModels : INotifyPropertyChanged
+    {
+        //событие изменения свойства
+        public event PropertyChangedEventHandler PropertyChanged; 
```
сначала приватные поля
потом публичные

Commit:
В класс City добавлен конструктор.
Другие мелкие изменения.

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,76 @@
+        public string CityName
+        {
+            private get { return _cityName; }
+            set
```
приватный сет, а не гет
у тебя ошибки не возникает в чекбоксе?
ты только что запретил всем, кроме этого класса брать значение поля _cityName

Antonybae replied:
сет кстати можно вообще удалить

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,76 @@
+            if (name == null)
+                throw new ArgumentException();
+            else
+                CityName = name;
```
сохраняй значение не через свойство, а через поле
так быстрее

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,46 @@
+    //класс RelayCommand, реализующий интерфейс ICommand
+    public class RelayCommand : ICommand
+    {
+        private Action _execute;
```
пускай будет private readonly
это не позволит менять это поле за пределами конструктора

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,46 @@
+    public class RelayCommand : ICommand
+    {
+        private Action _execute;
+        private Predicate<object> _canExecute;
```
тоже private readonly

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        public RelayCommand(Action execute, Predicate<object> canExecute = null)
+        {
+            _execute = execute ?? throw new ArgumentException();
+            if (canExecute != null)
```
тоже через оператор ??
как выше

LilySweetCat commented on Task1/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            if (_canExecute != null)
+                return _canExecute(parameter);
+            return true;
```
прикольно получается
если может исполняться = null, то верни, что может исполняться
а вот если может исполняться != null, то тогда верни значение функции.
не думаешь, что тут где то ошибка?

LilySweetCat reviewed (commented):
:((

Commit:
Мелкие исправления

LilySweetCat commented on Task1/Model/City.cs:
```diff
@@ -0,0 +1,76 @@
+
+        //конструктор класса "Город"
+        public City(string name)
+        {
```
??

Commit:
Поправлен конструктор класса City на более удобную запись
